kroxylicious/src/main/java/io/kroxylicious/proxy/filter/NetFilter.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.filter;
+
+import java.net.SocketAddress;
+
+/**
+ * Abstracts some policy/logic for how an upstream connection for a given client connection
+ * is made.
+ */
+public interface NetFilter {
+
+    /**
+     * Determine the upstream cluster to connect to based on the information
+     * provided by the given {@code context},
+     * by invoking {@link NetFilterContext#initiateConnect(String, int, KrpcFilter[])}.
+     * @param context The context.
+     */
+    void selectServer(NetFilterContext context);
```

Comment:
The general pattern we chose for the API between filters and kroxy was:

* The write of a filter writes some subclass that implements an interface
* The methods of the interface take a context object
* The context object exposes information to the implementation via methods (this allows use to expose more information in the future without the interface method signature changing, i.e. avoid breaking changes in the API).
* The context also exposes methods for interacting with the proxy. For protocol filters we did this because having a `onProduceRequest()` which returned the (possibly modified) `ProduceRequest` to forward to the next filter was too limiting. E.g. we wanted for filters to do things like not forwarding requests/responses right now and making internal requests and on receipt of a response would allow us to forward the original request/response. (On example of how this is practically useful is the Topic Encryption use case, where you receive a `FetchResponse` which has only a topic id and an encryption policy which depends on the topic name; using this mechanism allows us to make an internal `MetadataRequest` to find the topic with the given id, the decrypt the records in the `FetchResponse` and forward it).

What I'm doing here is simply sticking to the same pattern. I think it would be more confusing to filter implementers if there was an inconsistency between interfaces, or methods of the same plugin interface. But it also would allow us to implement things like rate limiting within the proxy, for example. e.g. the context could expose a method to schedule a delayed connection.

Reply:
I whole heartedly agree consistency is important! 


> The context object exposes information to the implementation via methods (this allows use to expose more information in the future without the interface method signature changing, i.e. avoid breaking changes in the API).

I don't really follow this point. To add additional information via the methods surely the API is changing at some level? Either in the method itself, the context object or the type returned by the method. I can see that it buys API stability at the `NetFilter` level so changes only affect a subset of filters.

The fact that we are trying to enforce implementation details via documentation still feels dubious to me. I don't think I'm objecting to the general shape of the API but more the specifics of what is encoded here.

Reply:
> To add additional information via the methods surely the API is changing at some level? 

I meant it's a backwards compatible change. Specifically, if Alice has a some `FooNetFilter` class compiled against `v1` of the NetFilter API then us adding a new method to the `NetFilterContext` to expose some additional information, or enable some new way of interacting with the proxy layer, cannot break that existing `NetFilter` implementation.

If `NetFilter` looked like:

```java
interface NetFilter {
    HostAndPort selectServer(String sniHost, String authorizedId, String clientHost, int clientPort);
}
```

then changes are more difficult. Sure we can use `default` methods, like this:

```java
interface NetFilter {
    HostAndPort selectServer(String sniHost, String authorizedId, String clientHost, int clientPort);
    default HostAndPort selectServer(String sniHost, String authorizedId, String clientHost, int clientPort, String someNewProperty) {
        selectServer(sniHost, authorizedId, clientHost, clientPort);
    }
}
```

but then:
* people using the new interface now have to implement _both_ the `abstract` and the `default` method
* over time we end up with several `default` methods, 

Similar problems with `HostAndPort`. 

> The fact that we are trying to enforce implementation details via documentation still feels dubious to me.

I don't know what exactly this is referring to. I _guess_ it's the requirement that `selectServer` calls back on `initiateConnect`? I don't think that's an "implementation detail", but it _is_ part of a documented contract. I think that's fine; contracts are rarely _fully_ specified by type declarations, you often need some documentation in addition.

Reply:
I was thinking a name like `onXXX()` or `afterYYY()` might be a better fit (both in terms of consistency with `KrpcFilter`, but also because it would convey _when_ this method was being called and bring fewer assumptions about how the filter implementation should behave). The problem is that there is no good term for `XXX` or `YYY` because exactly _when_ the net filter gets called depends a lot on the client and/or the authentication. For example it's not like it's `afterAuthentication` (because authentication isn't necessarily configured). So far `onClusterAccess` is the best I've come up with. But that doesn't quite do it for me either.

Reply:
Apparently I failed to submit the review last week 😭 

I've tried to demonstrate my thinking in https://github.com/SamBarker/kproxy/commit/d079c2249def1c8ea2d2a3913166d8e038bb3475 

I don't know that it really changes the API in any practical way, we still have the call by contract and awkward terminology. However I think it shifts the intent/scope/responsibilities of API around again. The `FrontendHandler` becomes responsible again for the actual connection to a backend, the filters responsibility narrows to providing just the connection details for that backend. 

I think the contract becomes more obvious. The frontend handler calls `selectServer` and implementations call `selectedServer` (we can iterate on the naming obviously). I'm still a bit sceptical as to the need for contract rather than a return value, however I buy the consistency argument.

I've also tried https://github.com/SamBarker/kproxy/commit/dee61fa9fff49aaccd341cbd526992a00aac104e which violates the consistency point too I guess, but possibly makes the relationship more explicit.